src/triangulum/config.clj

```diff
@@ -1,80 +1,95 @@
+(s/def ::namespaced-config (s/merge ::config-namespaced/server
+                                    ::config-namespaced/app
+                                    ::config-namespaced/database
+                                    ::config-namespaced/mail
+                                    ::config-namespaced/https))
+
+;;; Private vars
 
 (def ^:private config-file  (atom "config.edn"))
 (def ^:private config-cache (atom nil))
+(def ^:private ns->un-mapping
+  "Maps namespaced keys namesapces to their unnamespaced counterparts."
+  {:views :app, :email :mail})
+(def ^:private un->ns-mapping
+  "Maps unnamespaced keys namespaces to their namespaced counterparts."
+  (reverse-map ns->un-mapping))
+
 
 ;;; Helper Fns
 
 (defn- wrap-throw [& strs]
   (-> (apply str strs)
-      (ex-info {})
-      (throw)))
+      (init-throw)))
+
+(defn- namespaced-key?
+  "Returns true if the given key has a namespace, otherwise false."
+  [k]
+  (namespace k))
 
 (defn- read-config [file]
   (if (.exists (io/file file))
-    (let [example-config (-> (slurp *default-file*) (edn/read-string))
-          config         (-> (slurp file) (edn/read-string))
-          missing-keys   (find-missing-keys example-config config)]
-      (cond
-        (seq missing-keys)
-        (wrap-throw "Error: The following keys from config.default.edn are missing from:"
-                    file
-                    "\n"
-                    (str/join "', '" missing-keys))
-
-        (not (s/valid? ::config config))
-        (do (println "Error: Invalid config file:" file)
-            (s/explain ::config config)
-            (flush)
-            (wrap-throw ""))
+    (if-let [config (nil-on-error (edn/read-string (slurp file)))]
+      (let [valid-nested-config? (s/valid? ::nested-config config)
+            valid-namespaced-config? (s/valid? ::namespaced-config config)]
+        (cond (or valid-nested-config?
+                  valid-namespaced-config?)
+              config
 
-        :else
-        config))
-    (wrap-throw "Error: Cannot find file" file)))
+              (every? namespaced-key? (keys config))
+              (wrap-throw "Error: Config file " file " failed spec check:\n" (s/explain-str ::namespaced-config config))
+
+              :else
+              (wrap-throw "Error: Config file " file " failed spec check:\n" (s/explain-str ::nested-config config))))
+      (wrap-throw "Error: Config file " file " does not contain valid EDN."))
+    (wrap-throw "Error: Cannot find config file " file ".")))
 
 (defn- cache-config []
   (or @config-cache
       (reset! config-cache (read-config @config-file))))
 
+(defn- get-mapped-key-ns
+  "Given a namespaced key, returns the corresponding unnamespaced key."
+  [ns-key]
+  (let [new-ns (-> ns-key
+                   namespace
+                   (clojure.string/split #"\.")
+                   second
+                   keyword)]
+    (or (get ns->un-mapping new-ns) new-ns)))
```

Comment:
We can remove the `or` function call here, passing `new-ns` as the default value to the function `get`.

e.g: `(get ns->un-mapping new-ns new-ns)`